README
Driver documentation for yealink phones and ATA adapter.

0. Status
~~~~~~~~~
Phone models:
USB-P1K	USB 1.1 hand phone with LCD, keypad and ringer.
USB-P4K	USB 1.1 speaker phone desktop model.
USB-B2K	USB 1.1 telbox, ATA adapter for PSTN.

For manufacturer documentation see http://www.yealink.com

The development branch is maintained at:
http://cvs.savannah.gnu.org/viewcvs/usbb2k-api/yealink-module/


Matrix of supported features:
function	API	P1K	P4K	B2K
-------------------------------------------
audio playback	alsa 	ok	ok	ok
audio record	alsa	ok	ok	ok
keyboard	input	ok	ok	WIP
hookflash	input		ok	WIP
LCD		sysfs	ok	ok
LED		sysfs	ok
DIALTONE	sysfs	ok	ok	ok
RINGTONE	sysfs	ok		WIP
BACKLIGHT	sysfs		ok
SPEAKER		sysfs		ok
PSTN		sysfs			ok
LED_USB		sysfs			WIP

Notes:
P4K: Ring tones are implemented by switching on the SPEAKER
and sending a ring tone pcm via the dsp interface.


0.1 User space applications
~~~~~~~~~~~~~~~~~~~~~~~~~~~

0.1.1 YeaPhone
~~~~~~~~~~~~~~
Thomas Reitmayr has released YeaPhone, a great addition to linphone, this
will turn your P1K into a full blown SIP phone. There are even packages
for the Linksys NSLU2 for an energy efficient and fanless solution!

Please check out: http://www.devbase.at/voip/yeaphone.php


1. Compilation (stand alone version)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Currently only kernel 2.6.x.y versions are supported.
In order to build the yealink.ko module do

  make

If you encounter problems please check if in the MAKE_OPTS variable
in the Makefile is pointing to the location where your kernel sources
are located, default /usr/src/linux.


1.1 Troubleshooting
~~~~~~~~~~~~~~~~~~~
Q: Module yealink compiled and installed without any problem but phone
   is not initialized and does not react to any actions.
A: If you see something like:
   hiddev0: USB HID v1.00 Device [Yealink Network Technology Ltd. VOIP USB Phone
   in dmesg, it means that the hid driver has grabbed the device first. Try to
   load module yealink before any other usb hid driver. Please see the
   instructions provided by your distribution on module configuration.
   UPDATE: since kernel version 2.6.18 the HID driver is patched in order to
   prevent grabbing the yealink device.

Q: Phone is working now (displays version and accepts keypad input) but I can't
   find the sysfs files.
A: The sysfs files are located on the particular usb endpoint. On most
   distributions you can do: "find /sys/ -name get_icons" for a hint.


2. keyboard features
~~~~~~~~~~~~~~~~~~~~
Keyboard events are processes trough the input layer. A user space application
may issue the EVIOCGRAB ioctl() on the corresponding /dev/input/eventX device
to prevent the key codes from the event device to go to the rest of the system.

The current mappings from scancode to input event are provided by the
map_p1k_to_key, map_p4k_to_key, ... functions.
See yealink.c for a description.


2.1 P1K keyboard layout
~~~~~~~~~~~~~~~~~~~~~~~

   Physical USB-P1K button layout	input events


              up			     up
        IN           OUT		left,	right
             down			    down

      pickup   C    hangup		enter, backspace, escape
        1      2      3			1, 2, 3
        4      5      6			4, 5, 6,
        7      8      9			7, 8, 9,
        *      0      #			*, 0, #,

  The "up" and "down" keys, are symbolized by arrows on the button.
  The "pickup" and "hangup" keys are symbolized by a green and red phone
  on the button.


2.2 Hookflash
~~~~~~~~~~~~~
The P4K generates KEY_PHONE up and down events when the phone taken off
or on -hook.


3. LCD features
~~~~~~~~~~~~~~~
The LCD is divided and organized as a 3 line display:

    |[]   [][]   [][]   [][]   in   |[][]
    |[] M [][] D [][] : [][]   out  |[][]
                              store

    NEW REP         SU MO TU WE TH FR SA

    [] [] [] [] [] [] [] [] [] [] [] []
    [] [] [] [] [] [] [] [] [] [] [] []


Line 1	Format (see below)	: 18.e8.M8.88...188
	Icon names		:   M  D  :  IN OUT STORE
Line 2  Format			: .........
	Icon name		: NEW REP SU MO TU WE TH FR SA
Line 3  Format			: 888888888888


Format description:
  From a user space perspective the world is separated in "digits" and "icons".
  A digit can have a character set, an icon can only be ON or OFF.

  Format specifier
    '8' :  Generic 7 segment digit with individual addressable segments

    Reduced capability 7 segm digit, when segments are hard wired together.
    '1' : 2 segment digit only able to produce a 1.
    'e' : Most significant day of the month digit,
          able to produce at least 1 2 3.
    'M' : Most significant minute digit,
          able to produce at least 0 1 2 3 4 5.

    Icons or pictograms:
    '.' : For example like AM, PM, SU, a 'dot' .. or other single segment
	  elements.


4. Driver usage
~~~~~~~~~~~~~~~
For userland the following interfaces are available using the sysfs interface:
  /sys/.../
           line1	Read/Write, lcd line1
           line2	Read/Write, lcd line2
           line3	Read/Write, lcd line3

	   get_icons    Read, returns a set of available icons.
	   hide_icon    Write, hide the element by writing the icon name.
	   show_icon    Write, display the element by writing the icon name.

	   map_seg7	Read/Write, the 7 segments char set, common for all
	   		yealink phones. (see map_to_7segment.h)

	   ringtone	Write, upload binary representation of a ringtone,
	   		see yealink.c. status EXPERIMENTAL due to potential
			races between async. and sync usb calls.

	   model	Read, returns available phone models with the currently
	   		selected one.
			Write, switches of phone model in order to get the
			correct key mapping for example.

  module parameters:

	default_model	Initializes the current model type, see /sys/../model.
			For example:

			insmod yealink.ko default_model=P1K
			insmod yealink.ko default_model=P4K

			This is a make shift solution until we know how to query
			the phone properly for this.


4.1 lineX
~~~~~~~~~
Reading /sys/../lineX will return the format string with its current value:

  Example:
  cat ./line3
  888888888888
  Linux Rocks!

Writing to /sys/../lineX will set the corresponding LCD line.
 - Excess characters are ignored.
 - If less characters are written than allowed, the remaining digits are
   unchanged.
 - The tab '\t'and '\n' char does not overwrite the original content.
 - Writing a space to an icon will always hide its content.

  Example:
  date +"%m.%e.%k:%M"  | sed 's/^0/ /' > ./line1

  Will update the LCD with the current date & time.


4.2 get_icons
~~~~~~~~~~~~~
Reading will return all available icon names and its current settings:

  cat ./get_icons
  on M
  on D
  on :
     IN
     OUT
     STORE
     NEW
     REP
     SU
     MO
     TU
     WE
     TH
     FR
     SA
     LED
     DIALTONE
     RINGTONE


4.3 show/hide icons
~~~~~~~~~~~~~~~~~~~
Writing to these files will update the state of the icon.
Only one icon at a time can be updated.

If an icon is also on a ./lineX the corresponding value is
updated with the first letter of the icon.

  Example - light up the store icon:
  echo -n "STORE" > ./show_icon

  cat ./line1
  18.e8.M8.88...188
               S

  Example - sound the ring tone for 10 seconds:
  echo -n RINGTONE > /sys/..../show_icon
  sleep 10
  echo -n RINGTONE > /sys/..../hide_icon


4.4 model
~~~~~~~~~
Select and query phone model.

  Example - show available types (with P1K selected):
  cat /sys/.../model
  * P1K
    P4K

  Example - switch to P4K model:
  echo -n "P4K" > /sys/.../model


5. Sound features
~~~~~~~~~~~~~~~~~
Sound is supported by the ALSA driver: snd_usb_audio

One 16-bit channel with sample and playback rates of 8000 Hz is the practical
limit of the device.

  Example - recording test:
  arecord -v -d 10 -r 8000 -f S16_LE -t wav  foobar.wav

  Example - playback test:
  aplay foobar.wav


6. Credits & Acknowledgments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
See yealink.c


CHANGES
This log is discontinued, for further reference see the yealink.c header.

20060430:
  Toshiharu Ohno: - Fixed key event initialization bug.

20051230:
  - sync with kernel 2.6.16

r20051228:
  - More selective usb_device table, removal of usb_match function.

r0816 :
  - Folow sysfs API changes that whent in kernel 2.6.13

r0802 :
  - Is considered usable for kernel versions up to 2.6.12.x


map_to_7segment.h
/*
 * drivers/usb/input/map_to_7segment.h
 *
 * Copyright (c) 2005,2006 Henk Vergonet <Henk.Vergonet@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef MAP_TO_7SEGMENT_H
#define MAP_TO_7SEGMENT_H

/* This file provides translation primitives and tables for the conversion
 * of (ASCII) characters to a 7-segments notation.
 *
 * The 7 segment's wikipedia notation below is used as standard.
 * See: http://en.wikipedia.org/wiki/Seven_segment_display
 *
 * Notation:	+-a-+
 *		f   b
 *		+-g-+
 *		e   c
 *		+-d-+
 *
 * Usage:
 *
 *   Register a map variable, and fill it with a character set:
 * 	static SEG7_DEFAULT_MAP(map_seg7);
 *
 *
 *   Then use for conversion:
 *	seg7 = map_to_seg7(&map_seg7, some_char);
 *	...
 *
 * In device drivers it is recommended, if required, to make the char map
 * accessible via the sysfs interface using the following scheme:
 *
 * static ssize_t show_map(struct device *dev, char *buf) {
 *	memcpy(buf, &map_seg7, sizeof(map_seg7));
 *	return sizeof(map_seg7);
 * }
 * static ssize_t store_map(struct device *dev, const char *buf, size_t cnt) {
 *	if(cnt != sizeof(map_seg7))
 *		return -EINVAL;
 *	memcpy(&map_seg7, buf, cnt);
 *	return cnt;
 * }
 * static DEVICE_ATTR(map_seg7, PERMS_RW, show_map, store_map);
 *
 * History:
 * 2005-05-31	RFC linux-kernel@vger.kernel.org
 */
#include <linux/errno.h>


#define BIT_SEG7_A		0
#define BIT_SEG7_B		1
#define BIT_SEG7_C		2
#define BIT_SEG7_D		3
#define BIT_SEG7_E		4
#define BIT_SEG7_F		5
#define BIT_SEG7_G		6
#define BIT_SEG7_RESERVED	7

struct seg7_conversion_map {
	unsigned char	table[128];
};

static inline int map_to_seg7(struct seg7_conversion_map *map, int c)
{
	return c >= 0 && c < sizeof(map->table) ? map->table[c] : -EINVAL;
}

#define SEG7_CONVERSION_MAP(_name, _map)	\
	struct seg7_conversion_map _name = { .table = { _map } }

/*
 * It is recommended to use a facility that allows user space to redefine
 * custom character sets for LCD devices. Please use a sysfs interface
 * as described above.
 */
#define MAP_TO_SEG7_SYSFS_FILE	"map_seg7"

/*******************************************************************************
 * ASCII conversion table
 ******************************************************************************/

#define _SEG7(l,a,b,c,d,e,f,g)	\
      (	a<<BIT_SEG7_A |	b<<BIT_SEG7_B |	c<<BIT_SEG7_C |	d<<BIT_SEG7_D |	\
	e<<BIT_SEG7_E |	f<<BIT_SEG7_F |	g<<BIT_SEG7_G )

#define _MAP_0_32_ASCII_SEG7_NON_PRINTABLE	\
	0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,

#define _MAP_33_47_ASCII_SEG7_SYMBOL		\
 _SEG7('!',0,0,0,0,1,1,0), _SEG7('"',0,1,0,0,0,1,0), _SEG7('#',0,1,1,0,1,1,0),\
 _SEG7('$',1,0,1,1,0,1,1), _SEG7('%',0,0,1,0,0,1,0), _SEG7('&',1,0,1,1,1,1,1),\
 _SEG7('\'',0,0,0,0,0,1,0),_SEG7('(',1,0,0,1,1,1,0), _SEG7(')',1,1,1,1,0,0,0),\
 _SEG7('*',0,1,1,0,1,1,1), _SEG7('+',0,1,1,0,0,0,1), _SEG7(',',0,0,0,0,1,0,0),\
 _SEG7('-',0,0,0,0,0,0,1), _SEG7('.',0,0,0,0,1,0,0), _SEG7('/',0,1,0,0,1,0,1),

#define _MAP_48_57_ASCII_SEG7_NUMERIC		\
 _SEG7('0',1,1,1,1,1,1,0), _SEG7('1',0,1,1,0,0,0,0), _SEG7('2',1,1,0,1,1,0,1),\
 _SEG7('3',1,1,1,1,0,0,1), _SEG7('4',0,1,1,0,0,1,1), _SEG7('5',1,0,1,1,0,1,1),\
 _SEG7('6',1,0,1,1,1,1,1), _SEG7('7',1,1,1,0,0,0,0), _SEG7('8',1,1,1,1,1,1,1),\
 _SEG7('9',1,1,1,1,0,1,1),

#define _MAP_58_64_ASCII_SEG7_SYMBOL		\
 _SEG7(':',0,0,0,1,0,0,1), _SEG7(';',0,0,0,1,0,0,1), _SEG7('<',1,0,0,0,0,1,1),\
 _SEG7('=',0,0,0,1,0,0,1), _SEG7('>',1,1,0,0,0,0,1), _SEG7('?',1,1,1,0,0,1,0),\
 _SEG7('@',1,1,0,1,1,1,1),

#define _MAP_65_90_ASCII_SEG7_ALPHA_UPPR	\
 _SEG7('A',1,1,1,0,1,1,1), _SEG7('B',1,1,1,1,1,1,1), _SEG7('C',1,0,0,1,1,1,0),\
 _SEG7('D',1,1,1,1,1,1,0), _SEG7('E',1,0,0,1,1,1,1), _SEG7('F',1,0,0,0,1,1,1),\
 _SEG7('G',1,1,1,1,0,1,1), _SEG7('H',0,1,1,0,1,1,1), _SEG7('I',0,1,1,0,0,0,0),\
 _SEG7('J',0,1,1,1,0,0,0), _SEG7('K',0,1,1,0,1,1,1), _SEG7('L',0,0,0,1,1,1,0),\
 _SEG7('M',1,1,1,0,1,1,0), _SEG7('N',1,1,1,0,1,1,0), _SEG7('O',1,1,1,1,1,1,0),\
 _SEG7('P',1,1,0,0,1,1,1), _SEG7('Q',1,1,1,1,1,1,0), _SEG7('R',1,1,1,0,1,1,1),\
 _SEG7('S',1,0,1,1,0,1,1), _SEG7('T',0,0,0,1,1,1,1), _SEG7('U',0,1,1,1,1,1,0),\
 _SEG7('V',0,1,1,1,1,1,0), _SEG7('W',0,1,1,1,1,1,1), _SEG7('X',0,1,1,0,1,1,1),\
 _SEG7('Y',0,1,1,0,0,1,1), _SEG7('Z',1,1,0,1,1,0,1),

#define _MAP_91_96_ASCII_SEG7_SYMBOL		\
 _SEG7('[',1,0,0,1,1,1,0), _SEG7('\\',0,0,1,0,0,1,1),_SEG7(']',1,1,1,1,0,0,0),\
 _SEG7('^',1,1,0,0,0,1,0), _SEG7('_',0,0,0,1,0,0,0), _SEG7('`',0,1,0,0,0,0,0),

#define _MAP_97_122_ASCII_SEG7_ALPHA_LOWER	\
 _SEG7('A',1,1,1,0,1,1,1), _SEG7('b',0,0,1,1,1,1,1), _SEG7('c',0,0,0,1,1,0,1),\
 _SEG7('d',0,1,1,1,1,0,1), _SEG7('E',1,0,0,1,1,1,1), _SEG7('F',1,0,0,0,1,1,1),\
 _SEG7('G',1,1,1,1,0,1,1), _SEG7('h',0,0,1,0,1,1,1), _SEG7('i',0,0,1,0,0,0,0),\
 _SEG7('j',0,0,1,1,0,0,0), _SEG7('k',0,0,1,0,1,1,1), _SEG7('L',0,0,0,1,1,1,0),\
 _SEG7('M',1,1,1,0,1,1,0), _SEG7('n',0,0,1,0,1,0,1), _SEG7('o',0,0,1,1,1,0,1),\
 _SEG7('P',1,1,0,0,1,1,1), _SEG7('q',1,1,1,0,0,1,1), _SEG7('r',0,0,0,0,1,0,1),\
 _SEG7('S',1,0,1,1,0,1,1), _SEG7('T',0,0,0,1,1,1,1), _SEG7('u',0,0,1,1,1,0,0),\
 _SEG7('v',0,0,1,1,1,0,0), _SEG7('W',0,1,1,1,1,1,1), _SEG7('X',0,1,1,0,1,1,1),\
 _SEG7('y',0,1,1,1,0,1,1), _SEG7('Z',1,1,0,1,1,0,1),

#define _MAP_123_126_ASCII_SEG7_SYMBOL		\
 _SEG7('{',1,0,0,1,1,1,0), _SEG7('|',0,0,0,0,1,1,0), _SEG7('}',1,1,1,1,0,0,0),\
 _SEG7('~',1,0,0,0,0,0,0),

/* Maps */

/* This set tries to map as close as possible to the visible characteristics
 * of the ASCII symbol, lowercase and uppercase letters may differ in
 * presentation on the display.
 */
#define MAP_ASCII7SEG_ALPHANUM			\
	_MAP_0_32_ASCII_SEG7_NON_PRINTABLE	\
	_MAP_33_47_ASCII_SEG7_SYMBOL		\
	_MAP_48_57_ASCII_SEG7_NUMERIC		\
	_MAP_58_64_ASCII_SEG7_SYMBOL		\
	_MAP_65_90_ASCII_SEG7_ALPHA_UPPR	\
	_MAP_91_96_ASCII_SEG7_SYMBOL		\
	_MAP_97_122_ASCII_SEG7_ALPHA_LOWER	\
	_MAP_123_126_ASCII_SEG7_SYMBOL

/* This set tries to map as close as possible to the symbolic characteristics
 * of the ASCII character for maximum discrimination.
 * For now this means all alpha chars are in lower case representations.
 * (This for example facilitates the use of hex numbers with uppercase input.)
 */
#define MAP_ASCII7SEG_ALPHANUM_LC			\
	_MAP_0_32_ASCII_SEG7_NON_PRINTABLE	\
	_MAP_33_47_ASCII_SEG7_SYMBOL		\
	_MAP_48_57_ASCII_SEG7_NUMERIC		\
	_MAP_58_64_ASCII_SEG7_SYMBOL		\
	_MAP_97_122_ASCII_SEG7_ALPHA_LOWER	\
	_MAP_91_96_ASCII_SEG7_SYMBOL		\
	_MAP_97_122_ASCII_SEG7_ALPHA_LOWER	\
	_MAP_123_126_ASCII_SEG7_SYMBOL

#define SEG7_DEFAULT_MAP(_name)		\
	SEG7_CONVERSION_MAP(_name,MAP_ASCII7SEG_ALPHANUM)

#endif	/* MAP_TO_7SEGMENT_H */



yealink.h
/*
 * drivers/usb/input/yealink.h
 *
 * Copyright (c) 2005,2006 Henk Vergonet <Henk.Vergonet@gmail.com>
 *               2008      Thomas Reitmayr <treitmayr@devbase.at>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
#ifndef INPUT_YEALINK_H
#define INPUT_YEALINK_H

/* Using the control channel on interface 3 various aspects of the phone
 * can be controlled like LCD, LED, dialtone and the ringtone.
 */

/* "Generation 1" models use 16 byte packets */

struct yld_ctl_packet_g1 {
	u8	cmd;		/* command code, see below */
	u8	size;		/* 1-11, size of used data bytes. */
	u16	offset;		/* internal packet offset */
	u8	data[11];
	s8	sum;		/* negative sum of 15 preceding bytes */
} __attribute__ ((packed));

#define USB_PKT_LEN_G1	sizeof(struct yld_ctl_packet_g1)

/* "Generation 2" models use 8 byte packets */

struct yld_ctl_packet_g2 {
	u8	cmd;		/* command code, see below */
	u8	data[6];
	s8	sum;		/* negative sum of 7 preceding bytes */
} __attribute__ ((packed));

#define USB_PKT_LEN_G2	sizeof(struct yld_ctl_packet_g2)

union yld_ctl_packet {
	u8	cmd;		/* command code is always the first byte */
	struct	yld_ctl_packet_g1 g1;
	struct	yld_ctl_packet_g2 g2;
};

enum yld_ctl_protocols {
  yld_ctl_protocol_g1,
  yld_ctl_protocol_g2
};

/* version ranges determined according to
 * http://www.devbase.at/svn/view.cgi/yealink-logs/version-numbers.txt?root=experimental
 */
#define YLD_IS_P1K(v)	((v) >= 0x0100 && (v) <= 0x01ff)
#define YLD_IS_P4K(v)	((v) >= 0x0230 && (v) <= 0x02ff)
#define YLD_IS_B2K(v)	(((v) >= 0x0520 && (v) <= 0x053f) || \
			 ((v) >= 0x0570 && (v) <= 0x058f))
#define YLD_IS_B3K(v)	((v) >= 0x0540 && (v) <= 0x056f)

/* The following yld_ctl_packet's are available: */

/* Init registers
 *
 * cmd		0x8e
 * size		10
 * offset	0
 * data		0,0,0,0....
 */
#define CMD_INIT		0x8e

/* Read version
 *
 * cmd		0x87
 * size		2
 * offset	0
 * data		0,0
 */
#define CMD_VERSION		0x87

/* Request scan of attached PSTN
 *
 * models       B2K (not B3K!)
 * cmd          0x8d
 * size         1
 * offset       0
 * data[0]      on return: 
 *                bit 0: 0 .. PSTN ring off or ring pause
                         1 .. PSTN ring on
 *                bit 1: 0 .. on hook
 *                       1 .. off hook
 * Note: Bit 0 turns on and off along with the PSTN ring signal. Then end
 *       of the ring sequence (other party hung up) is detected when bit 0
 *       stays 0 for a few seconds.
 *       If not already in PSTN mode, SW should switch to PSTN immediately
 *       when detecting the first ring signal (via CMD_PSTN_SWITCH).
 */
#define CMD_HANDSET	0x8d

/* Request key scan
 *
 * cmd		0x80
 * size		1
 * offset	0
 * data[0]	on return returns the key event sequence, if it changes
 *		there's a new key pressed.
 */
#define CMD_KEYPRESS		0x80

/* Request scancode
 *
 * cmd		0x81
 * size		1
 * offset	key number [0-1f]
 * data[0]	on return returns the scancode
 */
#define CMD_SCANCODE		0x81

/* Request Hook scan
 *
 * models	P3K, P4K, V1K
 * cmd		0x8b
 * size		1
 * offset	0
 * data[0]	on return: bit 4: 1 .. on hook / 0 .. off hook
 */
#define	CMD_HOOKPRESS		0x8b

/* Set LCD
 *
 * cmd		0x04
 * size		1-11
 * offset	0-23
 * data		segment bits
 */
#define CMD_LCD			0x04

/* Set led
 *
 * models       P1K, P1KH, B2K, B3K
 * cmd		0x05
 * size		1
 * offset	0
 * data[0]	1 OFF / 0 ON (P1K, P1KH)
 * data[0,1]	"00 ff" OFF / "ff 00" ON (B2K, B3K)
 */
#define CMD_LED			0x05

/* Set ringtone volume
 *
 * models       P1K, P1KH
 * cmd		0x11
 * size		1
 * offset	0
 * data[0]	0-0xff  volume
 */
#define CMD_RING_VOLUME		0x11

/* Turn on the speaker for ringing and hands-free operation
 *
 * models       P3K, P4K, V1K
 * cmd		0x0c
 * size		1
 * offset	0
 * data[0]	0 .. off, 1 .. on
 * Note: The P4K has no buzzer, so the speaker and an audio ring tone
 *       have to be used for ringing.
 */
#define CMD_SPEAKER             0x0c

/* Set ringtone notes
 *
 * models	P1K, P1KH
 * cmd		0x02
 * size		1-11
 * offset	0->
 * data		P1K:  binary representation LE16(-freq), LE16(duration) ....
 *		P1KH: only 1 byte for each -freq, duration
 * Note: Current investigations with the P1KH suggest that the P1KH requires
 *       the ring notes to be downloaded each time before turning on the
 *       ringtone.
 */
#define CMD_RING_NOTE		0x02

/* Sound ringtone via the speaker on the back
 *
 * models	P1K, P1KH
 * cmd		0x03
 * size		1
 * offset	0
 * data[0]	0 OFF / 0x24 ON (P1K)
 *              0 OFF / 0xff ON (P1KH)
 * Note: At least for the P1KH this seems to be a loop counter which is
 *       decremented with each sequence of ring notes. Therefore ringing will
 *       stop after the ring notes were played data[0] times.
 */
#define CMD_RINGTONE		0x03

/* Sound dial tone via the ear speaker
 *
 * models       B2K, B3K, P3K, P4K, V1K
 * cmd		0x09
 * size		1
 * offset	0
 * data[0]	0 OFF / 1 ON
 */
#define CMD_DIALTONE		0x09

/* LCD backlight control
 *
 * models	P3K, P4K, V1K
 * cmd		0x12
 * size		1
 * offset	0
 * data[0]	0 OFF / 1 ON
 */
#define CMD_LCD_BACKLIGHT	0x12

/* B2K Ring control
 *
 * models	B2K, B3K
 * cmd		0x01
 * size		1
 * offset	0
 * data[0]	0 OFF / 1 ON
 * Note: For a periodic ring tone this bit has to be turned on and off.
 */
#define CMD_B2K_RING		0x01

/* B2K PSTN/USB switch
 *
 * models	B2K, B3K
 * cmd		0x0e
 * size		1
 * offset	0
 * data[0]	0 USB / 1 PSTN
 * Note: It is recommended to turn this on (PSTN) when the program shuts down.
 *       For the B3K if call forwarding is enabled, this has to be 0.
 */
#define CMD_PSTN_SWITCH		0x0e

/* Overload state
 *
 * models	?
 * state	0xfd
 * Note: This state seems to be replied whenever the received packet had an
 *       invalid checksum or some other internal error occurred in the phone.
 */
#define STATE_BAD_PKT		0xfd


#endif /* INPUT_YEALINK_H */


#if defined(_SEG) && defined(_PIC)
/* This table maps the LCD segments onto individual bit positions in the
 * yld_status struct.
 */

/* LCD, each segment must be driven seperately.
 *
 * Layout:
 *
 *   |[]   [][]   [][]   [][]   in   |[][]
 *   |[] M [][] D [][] : [][]   out  |[][]
 *                             store
 *
 *    NEW REP         SU MO TU WE TH FR SA
 *
 *    [] [] [] [] [] [] [] [] [] [] [] []
 *    [] [] [] [] [] [] [] [] [] [] [] []
 */

/* Line 1
 *	Format		: 18.e8.M8.88...188
 *	Icon names	: M D : IN OUT STORE
 */
#define LCD_LINE1_OFFSET	0
#define LCD_LINE1_SIZE		17

/* Note: first g then f =>			       !      !      */
/* _SEG(    type    a      b      c      d      e      g      f   )  */
	_SEG('1',  0,0 , 22,2 , 22,2 ,  0,0 ,  0,0 ,  0,0 ,  0,0	),
	_SEG('8', 20,1 , 20,2 , 20,4 , 20,8 , 21,4 , 21,2 , 21,1	),
	_PIC('.', 22,1 , "M"						),
	_SEG('e', 18,1 , 18,2 , 18,4 , 18,1 , 19,2 , 18,1 , 19,1	),
	_SEG('8', 16,1 , 16,2 , 16,4 , 16,8 , 17,4 , 17,2 , 17,1	),
	_PIC('.', 15,8 , "D"						),
	_SEG('M', 14,1 , 14,2 , 14,4 , 14,1 , 15,4 , 15,2 , 15,1	),
	_SEG('8', 12,1 , 12,2 , 12,4 , 12,8 , 13,4 , 13,2 , 13,1	),
	_PIC('.', 11,8 , ":"						),
	_SEG('8', 10,1 , 10,2 , 10,4 , 10,8 , 11,4 , 11,2 , 11,1	),
	_SEG('8',  8,1 ,  8,2 ,  8,4 ,  8,8 ,  9,4 ,  9,2 ,  9,1	),
	_PIC('.',  7,1 , "IN"						),
	_PIC('.',  7,2 , "OUT"						),
	_PIC('.',  7,4 , "STORE"					),
	_SEG('1',  0,0 ,  5,1 ,  5,1 ,  0,0 ,  0,0 ,  0,0 ,  0,0	),
	_SEG('8',  4,1 ,  4,2 ,  4,4 ,  4,8 ,  5,8 ,  5,4 ,  5,2	),
	_SEG('8',  2,1 ,  2,2 ,  2,4 ,  2,8 ,  3,4 ,  3,2 ,  3,1	),

/* Line 2
 *	Format		: .........
 *	Pict. name	: NEW REP SU MO TU WE TH FR SA
 */
#define LCD_LINE2_OFFSET	LCD_LINE1_OFFSET + LCD_LINE1_SIZE
#define LCD_LINE2_SIZE		9

	_PIC('.', 23,2 , "NEW"	),
	_PIC('.', 23,4 , "REP"	),
	_PIC('.',  1,8 , "SU"	),
	_PIC('.',  1,4 , "MO"	),
	_PIC('.',  1,2 , "TU"	),
	_PIC('.',  1,1 , "WE"	),
	_PIC('.',  0,1 , "TH"	),
	_PIC('.',  0,2 , "FR"	),
	_PIC('.',  0,4 , "SA"	),

/* Line 3
 *	Format		: 888888888888
 */
#define LCD_LINE3_OFFSET	LCD_LINE2_OFFSET + LCD_LINE2_SIZE
#define LCD_LINE3_SIZE		12

	_SEG('8', 22,16, 22,32, 22,64, 22,128, 23,128, 23,64, 23,32  ),
	_SEG('8', 20,16, 20,32, 20,64, 20,128, 21,128, 21,64, 21,32  ),
	_SEG('8', 18,16, 18,32, 18,64, 18,128, 19,128, 19,64, 19,32  ),
	_SEG('8', 16,16, 16,32, 16,64, 16,128, 17,128, 17,64, 17,32  ),
	_SEG('8', 14,16, 14,32, 14,64, 14,128, 15,128, 15,64, 15,32  ),
	_SEG('8', 12,16, 12,32, 12,64, 12,128, 13,128, 13,64, 13,32  ),
	_SEG('8', 10,16, 10,32, 10,64, 10,128, 11,128, 11,64, 11,32  ),
	_SEG('8',  8,16,  8,32,  8,64,  8,128,  9,128,  9,64,  9,32  ),
	_SEG('8',  6,16,  6,32,  6,64,  6,128,  7,128,  7,64,  7,32  ),
	_SEG('8',  4,16,  4,32,  4,64,  4,128,  5,128,  5,64,  5,32  ),
	_SEG('8',  2,16,  2,32,  2,64,  2,128,  3,128,  3,64,  3,32  ),
	_SEG('8',  0,16,  0,32,  0,64,  0,128,  1,128,  1,64,  1,32  ),

/* Line 4
 *
 * The LED, DIALTONE and RINGTONE are implemented as icons and use the same
 * sysfs interface.
 */
#define LCD_LINE4_OFFSET	LCD_LINE3_OFFSET + LCD_LINE3_SIZE

	_PIC('.', offsetof(struct yld_status, led) -
		  offsetof(struct yld_status, lcd)	, 0x01, "LED"	   ),
	_PIC('.', offsetof(struct yld_status, dialtone) -
		  offsetof(struct yld_status, lcd)	, 0x01, "DIALTONE" ),
	_PIC('.', offsetof(struct yld_status, ringtone) -
		  offsetof(struct yld_status, lcd)	, 0x01, "RINGTONE" ),
	/* P4K specific: */
	_PIC('.', offsetof(struct yld_status, backlight) -
		  offsetof(struct yld_status, lcd)	, 0x01, "BACKLIGHT"),
	_PIC('.', offsetof(struct yld_status, speaker) -
		  offsetof(struct yld_status, lcd)	, 0x01, "SPEAKER"  ),
	/* B2K specific: */
	_PIC('.', offsetof(struct yld_status, pstn) -
		  offsetof(struct yld_status, lcd)	, 0x01, "PSTN"	   ),

#undef _SEG
#undef _PIC
#endif /* _SEG && _PIC */
